pkg/controller/job/job_controller.go

```diff
@@ -1292,15 +1306,25 @@ func getStatus(job *batch.Job, pods []*v1.Pod, uncounted *uncountedTerminatedPod
 		return p.Status.Phase == v1.PodSucceeded
 	}))
 	failed += int32(countValidPodsWithFilter(job, pods, uncounted.Failed(), expectedRmFinalizers, func(p *v1.Pod) bool {
-		if p.Status.Phase == v1.PodFailed {
-			return true
+		isFailed := p.Status.Phase == v1.PodFailed
+
+		if !isFailed {
+			// When tracking with finalizers: counting deleted Pods as failures to
+			// account for orphan Pods that never have a chance to reach the Failed
+			// phase.
+			isFailed = uncounted != nil && p.DeletionTimestamp != nil && p.Status.Phase != v1.PodSucceeded
+		}
+		if isFailed && feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) {
+			terminationMessage, shouldCountFailed := matchPodFailurePolicy(job, p)
+			if jobTerminationMessage == nil && terminationMessage != nil {
+				jobTerminationMessage = terminationMessage
+			} else if !shouldCountFailed {
+				isFailed = false
+			}
 		}
-		// When tracking with finalizers: counting deleted Pods as failures to
-		// account for orphan Pods that never have a chance to reach the Failed
-		// phase.
-		return uncounted != nil && p.DeletionTimestamp != nil && p.Status.Phase != v1.PodSucceeded
+		return isFailed
 	}))
-	return succeeded, failed
+	return
```

Comment:
is this line necessary?

Reply:
Yes, without it I get a compilation error "MissingReturn". Alternatively, we can have:
```golang
return succeeded, failed, jobTerminationMessage
```
but this is not necessary as the function's result type uses named parameters: https://go.dev/ref/spec#Return_statements